Tools/scripts/docker_build.sh

```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+if [ $# -ne 1 ]; then
+    echo "No board supplied, run as ./docker_build.sh <board name> or ./docker_build.sh list"
+    exit 1
+fi
+
+BOARD=$1
+
+cd "$(git rev-parse --show-toplevel)"
+git submodule update --init --recursive
```

Comment:
we don't want this by default

Reply:
This PR is meant to enable one-step building of firmware for people who don't know about development. This way, all they need is Docker and they can build the firmware. Taking steps out of the script defeats is purpose, but feel free to close this PR if you don't want a one-step builder.

Reply:
updating the submodule is part of our instructions to setup the repository correctly, so that isn't an issue.

The main problem by hidding the submodule update there is that you hide it and then people won't understand why it fails. Moreover it will also prevent to work on some submodule as it will always erase the changes.

Reply:
> updating the submodule is part of our instructions to setup the repository correctly, so that isn't an issue.

Certainly, but then it's not "one-step". Having a "few steps" way is fine, and my main problem was that the PATH wasn't getting set correctly so the Docker image would only build if you entered the bash shell inside the container, but I also wanted a one-step way to build.

> The main problem by hidding the submodule update there is that you hide it and then people won't understand why it fails. Moreover it will also prevent to work on some submodule as it will always erase the changes.

I agree, but the users who know what "submodule" (or even "git") means wouldn't need to use this script, or would better be able to debug it if it breaks. This is specifically so the instructions can say "Run these commands: 'git clone; ./docker_build.sh' and the firmware will be in the `build/` directory".

It's fine if you think that, for example, novices shouldn't be trying to build, but currently the instructions are basically 3-4 steps. To me, reducing that number to 1 is worth it, 2 kind of isn't, so I'd be perfectly okay if we didn't merge this in the main repo and I could post a note somewhere with it, for new users.

I can also remove the submodule update, if you think that's worthwhile, but then it's just a script that contains two Docker commands, and might as well be `docker run bash -c "waf configure; waf build"`.

Reply:
I'm really a bit confused as to what happened here.  My understanding is that this was a stand-alone script which would let people get started developing ardupilot more easily than they currently do.  I'm a bit disappointed this was simply closed without a defining reason as to why it was a bad idea.

Note that in regards to the submodules - we should have patches in master which mean that they are magically initialised on first build - so it really ought not to have been an issue here.

Reply:
I closed it because, on one hand, there didn't seem to be much want for such a one-step script, and, on the other, it "works for me" and I didn't want to spend however many hours trying to figure out how to make bash load the appropriate PATH (though I'm mystified as to why it doesn't, it runs fine if you start bash in the container).